File: ResourceManager/Models/ResourceManagerDBContext.vb
Imports Microsoft.EntityFrameworkCore
Imports ResourceManager.Models.Configurations
Imports System.Linq
Imports System.Threading
Imports System.Threading.Tasks

Namespace ResourceManager.Models
    Public Class ResourceManagerDBContext
        Inherits DbContext
        Public Sub New(ByVal options As DbContextOptions)
            MyBase.New(options)
        End Sub

        Public Property Users As DbSet(Of User)
        Public Property Tags As DbSet(Of Tag)
        Public Property Roles As DbSet(Of Role)
        Public Property Permissions As DbSet(Of Permission)
        Public Property RolePermissions As DbSet(Of RolePermission)
        Public Property Passwords As DbSet(Of Password)
        Public Property PasswordTags As DbSet(Of PasswordTag)
        Public Property UserRoles As DbSet(Of UserRole)
        Public Property Books As DbSet(Of Book)
        Public Property BookTags As DbSet(Of BookTag)
        Public Property Articles As DbSet(Of Article)
        Public Property ArticleTags As DbSet(Of ArticleTag)
        Public Property Activities As DbSet(Of Activity)
        Public Property ActivityTags As DbSet(Of ActivityTag)
        Public Property Repositories As DbSet(Of Repository)
        Public Property RepositoryTags As DbSet(Of RepositoryTag)
        Public Property Notes As DbSet(Of Note)
        Public Property NoteTags As DbSet(Of NoteTag)
        Protected Overrides Sub OnModelCreating(ByVal modelBuilder As ModelBuilder)
            modelBuilder.ApplyConfiguration(New UserConfiguration())
            modelBuilder.ApplyConfiguration(New RoleConfiguration())
            modelBuilder.ApplyConfiguration(New TagConfiguration())
            modelBuilder.ApplyConfiguration(New PermissionConfiguration())
            modelBuilder.ApplyConfiguration(New RolePermissionConfiguration())
            modelBuilder.ApplyConfiguration(New UserRoleConfiguration())
            modelBuilder.ApplyConfiguration(New PasswordConfiguration())
            modelBuilder.ApplyConfiguration(New PasswordTagConfiguration())
            modelBuilder.ApplyConfiguration(New BookConfiguration())
            modelBuilder.ApplyConfiguration(New BookTagConfiguration())
            modelBuilder.ApplyConfiguration(New ActivityConfiguration())
            modelBuilder.ApplyConfiguration(New ActivityTagConfiguration())
            modelBuilder.ApplyConfiguration(New ArticleConfiguration())
            modelBuilder.ApplyConfiguration(New ArticleTagConfiguration())
            modelBuilder.ApplyConfiguration(New RepositoryConfiguration())
            modelBuilder.ApplyConfiguration(New RepositoryTagConfiguration())
            modelBuilder.ApplyConfiguration(New NoteConfiguration())
            modelBuilder.ApplyConfiguration(New NoteTagConfiguration())
        End Sub
        Public Overrides Function SaveChanges() As Integer
            Dim entries = MyBase.ChangeTracker.Entries().Where(Function(e) TypeOf e.Entity Is BaseEntity AndAlso (e.State = EntityState.Added OrElse e.State = EntityState.Modified))

            For Each entityEntry In entries
                CType(entityEntry.Entity, BaseEntity).ModifiedDate = Date.Now

                If entityEntry.State = EntityState.Added Then
                    CType(entityEntry.Entity, BaseEntity).CreatedDate = Date.Now
                End If
            Next

            Return MyBase.SaveChanges()
        End Function

        Public Overrides Function SaveChangesAsync(ByVal Optional cancellationToken As CancellationToken = Nothing) As Task(Of Integer)
            Dim entries = MyBase.ChangeTracker.Entries().Where(Function(e) TypeOf e.Entity Is BaseEntity AndAlso (e.State = EntityState.Added OrElse e.State = EntityState.Modified))

            For Each entityEntry In entries
                CType(entityEntry.Entity, BaseEntity).ModifiedDate = Date.Now

                If entityEntry.State = EntityState.Added Then
                    CType(entityEntry.Entity, BaseEntity).CreatedDate = Date.Now
                End If
            Next

            Return MyBase.SaveChangesAsync(cancellationToken)
        End Function
    End Class
End Namespace

File: ResourceManager/Models/ResourceManagerDBContext.cs
using Microsoft.EntityFrameworkCore;
using ResourceManager.Models.Configurations;
using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace ResourceManager.Models
{
    public class ResourceManagerDBContext : DbContext
    {
        public ResourceManagerDBContext(DbContextOptions options) : base(options) { }

        public DbSet<User> Users { get; set; }
        public DbSet<Tag> Tags { get; set; }
        public DbSet<Role> Roles { get; set; }
        public DbSet<Permission> Permissions { get; set; }
        public DbSet<RolePermission> RolePermissions { get; set; }
        public DbSet<Password> Passwords { get; set; }
        public DbSet<PasswordTag> PasswordTags { get; set; }
        public DbSet<UserRole> UserRoles { get; set; }
        public DbSet<Book> Books { get; set; }
        public DbSet<BookTag> BookTags { get; set; }
        public DbSet<Article> Articles { get; set; }
        public DbSet<ArticleTag> ArticleTags { get; set; }
        public DbSet<Activity> Activities { get; set; }
        public DbSet<ActivityTag> ActivityTags { get; set; }
        public DbSet<Repository> Repositories { get; set; }
        public DbSet<RepositoryTag> RepositoryTags { get; set; }
        public DbSet<Note> Notes { get; set; }
        public DbSet<NoteTag> NoteTags { get; set; }
        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.ApplyConfiguration(new UserConfiguration());
            modelBuilder.ApplyConfiguration(new RoleConfiguration());
            modelBuilder.ApplyConfiguration(new TagConfiguration());
            modelBuilder.ApplyConfiguration(new PermissionConfiguration());  
            modelBuilder.ApplyConfiguration(new RolePermissionConfiguration());
            modelBuilder.ApplyConfiguration(new UserRoleConfiguration());
            modelBuilder.ApplyConfiguration(new PasswordConfiguration());
            modelBuilder.ApplyConfiguration(new PasswordTagConfiguration());
            modelBuilder.ApplyConfiguration(new BookConfiguration());
            modelBuilder.ApplyConfiguration(new BookTagConfiguration());
            modelBuilder.ApplyConfiguration(new ActivityConfiguration());
            modelBuilder.ApplyConfiguration(new ActivityTagConfiguration());
            modelBuilder.ApplyConfiguration(new ArticleConfiguration());
            modelBuilder.ApplyConfiguration(new ArticleTagConfiguration());
            modelBuilder.ApplyConfiguration(new RepositoryConfiguration());
            modelBuilder.ApplyConfiguration(new RepositoryTagConfiguration());
            modelBuilder.ApplyConfiguration(new NoteConfiguration());
            modelBuilder.ApplyConfiguration(new NoteTagConfiguration());
        }
        public override int SaveChanges()
        {
            var entries = ChangeTracker
                .Entries()
                .Where(e => e.Entity is BaseEntity && (
                        e.State == EntityState.Added
                        || e.State == EntityState.Modified));

            foreach (var entityEntry in entries)
            {
                ((BaseEntity)entityEntry.Entity).ModifiedDate = DateTime.Now;

                if (entityEntry.State == EntityState.Added)
                {
                    ((BaseEntity)entityEntry.Entity).CreatedDate = DateTime.Now;
                }
            }

            return base.SaveChanges();
        }

        public override Task<int> SaveChangesAsync(CancellationToken cancellationToken = default)
        {
            var entries = ChangeTracker
                .Entries()
                .Where(e => e.Entity is BaseEntity && (
                        e.State == EntityState.Added
                        || e.State == EntityState.Modified));

            foreach (var entityEntry in entries)
            {
                ((BaseEntity)entityEntry.Entity).ModifiedDate = DateTime.Now;

                if (entityEntry.State == EntityState.Added)
                {
                    ((BaseEntity)entityEntry.Entity).CreatedDate = DateTime.Now;
                }
            }

            return base.SaveChangesAsync(cancellationToken);
        }
    }
}
